### PR TITLE
Diffractive imaging into dev

### PR DESCRIPTION
**Pull request by arthurmccray**
this is a big, messy PR. Unfortunately quite a bit of core has been changed in addition to the diffractive_imaging module, which was necessary for working with real-world experimental data. 

There has been quite a bit of code shared between diffractive_imaging and tomography, so after this PR we can pull from dev -> tomography to make that future PR easier as well. 

Hopefully after doing this we can go back to a proper system of only messing with one module at a time and smaller PRs.

### CODE REVIEW

**Review by gvarnavi — CHANGES_REQUESTED** [review]
Ok, finished first pass. I skipped all the splat/tomo modules, as-well as the tests. 

Good job with this re-structuring! I tried to leave detailed feedback, but the ptycho functionality is mostly good to go on my end. My main design comment is re: management of options and config files. I believe we should use [OmegaConf](https://omegaconf.readthedocs.io/en/2.3_branch/) which reads / writes various config patterns such as nested dictionaries and dot-files. This is particularly important for the hyperparameter optimization class.

I will spend the next couple of hours today implementing the following (since I likely won't have any more time this week):
- thin SVD version of probe orthogonalization
- torch-native ComplexProbe w/ soft apertures (no rolloff)
- give an OriginModel example

File: src/quantem/core/ml/cnn2d.py [hunk 1]
```diff
@@ -19,25 +19,28 @@ class CNN2d(nn.Module):
...
         self,
-        num_channels: int,  # (C, H, W) input shape same as output shape
+        in_channels: int,  # input channels (C_in, H, W)
+        out_channels: int | None = None,  # output channels (C_out, H, W)
         start_filters: int = 16,
         num_layers: int = 3,  # num_layers
         num_per_layer: int = 2,  # number conv per layer
-        use_skip_connections: bool = False,
-        dtype: torch.dtype = torch.complex64,
+        use_skip_connections: bool = True,
```
[the remark this answers is not in the document]
Comment (arthurmccray) [reply src/quantem/core/ml/cnn2d.py:27]: the default for skip_connections was previously "False" because it's called "CNN" rather than "UNET", but in 100% of cases the reconstructions converge faster and better with `use_skip_connections = True`, so i made it the default to be easier for users. 

Agreed re. changing defaults. We could potentially add warnings and/or even warn in advance (similar to other large packages). We have our global config as well so it would be pretty straightforward to disable them as well.

File: src/quantem/diffractive_imaging/ptychography_lite.py
[the remark this answers is not in the document]
Comment (arthurmccray) [reply src/quantem/diffractive_imaging/ptychography_lite.py:1]: That's a fair point, we discussed this a bit and thought "Lite" was the clearest. I'll raise it on the slack

File: src/quantem/diffractive_imaging/ptychography_opt.py
[the remark this answers is not in the document]
Comment (arthurmccray) [reply src/quantem/diffractive_imaging/ptychography_opt.py:1]: Agreed, though there's also an argument to be made that `ptychography_opt.py` shouldn't exist, and these bits should be wrapped into one of either `ptychography` or `ptychography_base`. In general though I do like being explicit with hyperparameter optimization, as there's plenty of other "optimizers" that we're using.

File: tests/conftest.py
[the remark this answers is not in the document]
Comment (arthurmccray) [reply tests/conftest.py:1]: Yup, the main thing I added was a "runslow" flag, cuz the ptycho tests are not fast. Default when calling pytest is to skip all the tests marked as slow, and only run them when called with the flag.

File: src/quantem/diffractive_imaging/object_models.py [hunk 2]
```diff
@@ -1,64 +1,123 @@
...
-    validate_array,
     validate_gt,
-    validate_np_len,
     validate_tensor,
 )
 from quantem.core.visualization import show_2d
+from quantem.diffractive_imaging.constraints import BaseConstraints
 from quantem.diffractive_imaging.ptycho_utils import sum_patches
 
+object_type = Literal["potential", "pure_phase", "complex"]
```
[the remark this answers is not in the document]
Comment (arthurmccray) [reply src/quantem/diffractive_imaging/object_models.py:27]: the use case is pretty much for DIPs, as it's quite a bit faster to use a real-valued CNN rather than complex. I'm not sure if there's an argument to be made for always using potential instead, but i have been using pure_phase in those cases.

File: src/quantem/diffractive_imaging/object_models.py [hunk 3]
```diff
@@ -303,38 +321,161 @@ def apply_constraints(
...
             obj_type=obj_type,
             rng=rng,
+            shape=shape,
+            _token=_token,
         )
-        self.constraints = self.DEFAULT_CONSTRAINTS.copy()
+        self._obj = nn.Parameter(torch.ones(1), requires_grad=True)
+
+    @classmethod
+    def from_uniform(
```
[the remark this answers is not in the document]
Comment (arthurmccray) [reply src/quantem/diffractive_imaging/object_models.py:449]: sure

File: src/quantem/diffractive_imaging/probe_models.py [hunk 4]
```diff
@@ -1,66 +1,99 @@
...
 
-class ProbeBase(AutoSerialize):
+
+class ProbeBase(nn.Module, RNGMixin, OptimizerMixin, AutoSerialize):
     DEFAULT_PROBE_PARAMS = {
         "energy": None,
         "defocus": None,
         "semiangle_cutoff": None,
         "rolloff": 2,
         "polar_parameters": {},
```
[the remark this answers is not in the document]
Comment (arthurmccray) [reply src/quantem/diffractive_imaging/probe_models.py:52]: sure

File: src/quantem/diffractive_imaging/probe_models.py [hunk 5]
```diff
@@ -318,125 +317,190 @@ def backward(self, *args, **kwargs):
...
-    # TODO write classmethods for from_params and from_array
-    # from_params should accept vacuum probe as well
+        return probe_model
 
     @property
     def probe(self) -> torch.Tensor:
         """get the full probe"""
-        return self.apply_constraints(self._probe)
-        # return self._probe
+        return self.apply_hard_constraints(self._probe)
```
[the remark this answers is not in the document]
Comment (arthurmccray) [reply src/quantem/diffractive_imaging/probe_models.py:525]: yeah but it's super cheap. It also is a really easy way of working around in place operations that can screw with backpropagation (especially when continuing a reconstruction for additional iterations), and it's consistent this way with how the DIP implementations work, where you have to apply hard constraints directly after getting the output of the model.

File: src/quantem/diffractive_imaging/ptychography_base.py [hunk 6]
```diff
@@ -377,13 +355,14 @@ def obj_padding_px(self, pad: np.ndarray | tuple[int, int]):
         if self._obj_padding_force_power2_level > 0:
             p2 = adjust_padding_power2(
```
[the remark this answers is not in the document]
Comment (arthurmccray) [reply src/quantem/diffractive_imaging/ptychography_base.py:356]: it's not about speed but is required for DIPs. if you're using a N layer deep CNN you need the object shape to be divisible by 2^N (except along the channel direction if doing a multislice recon with a 2D CNN)

File: src/quantem/diffractive_imaging/detector_models.py
[the remark this answers is not in the document]
Comment (arthurmccray) [reply src/quantem/diffractive_imaging/detector_models.py:1]: Sure, though I suggest that can wait until we have a usecase/need for it, unless you want to write it now.

File: pyproject.toml [hunk 7]
```diff
@@ -43,6 +41,11 @@ dependencies = [
   "rosettasciio>=0.8.0",
   "h5py>=3.13.0",
   "torch>=2.7.0",
+  "scikit-image>=0.25.2",
+  "torchvision>=0.22.1",
```
[the remark this answers is not in the document]
Comment (arthurmccray) [reply pyproject.toml:45]: tensorboard is used for logging of pytorch stuff too, and it allows live monitoring of reconstructions, loss components, etc. It's pretty neat actually, especially for big/slow/distributed reconstructions such as for tomo. I'm not sure how/if we want to do optional dependencies, but this could be optional.

File: src/quantem/diffractive_imaging/ptychography_opt.py [hunk 8]
```diff
@@ -0,0 +1,232 @@
...
+            for key, params in [
+                ("object", self.obj_model.optimizer_params),
+                ("probe", self.probe_model.optimizer_params),
+                ("dataset", self.dset.optimizer_params),
+            ]
+            if params
+        }
+
+    @optimizer_params.setter
+    def optimizer_params(self, d: dict) -> None:
```
[the remark this answers is not in the document]
Comment (gvarnavi) [reply src/quantem/diffractive_imaging/ptychography_opt.py:52]: I think the pattern would be something like:

# dictionaries
func(
  object_cfg= {"padding":10, "object_type":"complex"},
 probe_cfg= {"semiangle_cutoff":20}
)

# structured dataclasses
obj_cfg = ObjectConfig(padding=10,...)
probe_cfg = ProbeConfig(semiangle_cutoff...)
func(obj_cfg, probe_cfg)

# dot lists or dicts
kwargs = {'object.padding'=10, 'probe.semiangle_cutoff'=20}
func(**kwargs)

# from file
kwargs = OmegaConf.load('kwargs.yml')
func(**kwargs)

all of those work and internally get handled by OmegaConf.

File: src/quantem/diffractive_imaging/object_models.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (gvarnavi) [reply src/quantem/diffractive_imaging/object_models.py:27]: Does the code currently support running 10 iterations of potential, followed by 10 iterations of complex? If so, I don't think we lose anything by removing pure_phase

File: src/quantem/diffractive_imaging/ptychography.py [hunk 9]
```diff
@@ -362,19 +309,271 @@ def fourier_projection(self, measured_amplitudes, overlap_array):
...
+        Ptychography
+            Loaded ptychography object
+        """
+        # Load the base object without the dataset
+        ptycho = cls._recursive_load_from_path(path)
+
+        if not isinstance(ptycho, Ptychography):
+            raise ValueError("Loaded object is not a Ptychography object")
+
+        # If no dataset was provided, try to reload it from saved metadata
```
[the remark this answers is not in the document]
Comment (gvarnavi) [reply src/quantem/diffractive_imaging/ptychography.py:469]: To clarify, I think if you want to do the preprocessing for them that's fine, but I want them to pass the file location again. Files get moved around / renamed all the time, and it's especially confusing for EMPAD which uses the same filename

File: src/quantem/diffractive_imaging/ptychography_base.py [hunk 10]
```diff
@@ -1,14 +1,15 @@
-from pathlib import Path
-from typing import TYPE_CHECKING, Any, Literal, Sequence
+from typing import Any, Literal, Sequence, cast
+from warnings import warn
 
 import numpy as np
 import scipy.ndimage as ndi
+import torch
 
 import quantem.core.utils.array_funcs as arr
```
[the remark this answers is not in the document]
Comment (gvarnavi) [reply src/quantem/diffractive_imaging/ptychography_base.py:8]: See comment above, in another file you import it as arrf, and module names with arr are easily confused with array-like variable names
